### PR TITLE
DEC-796: User would like to remove/restore metadata fields

### PR DESCRIPTION
**Pull request by jgondron**
Why: User needs a way to remove and restore removed fields from their configuration, except for necessary fields like name.
How: 
  - Added an "active" property to all fields. 
  - Changed the configuration form to a list with a button that allows them to remove/restore each field. 
  - Added an immutable property list to each metadata field that allows us to restrict what properties a user can change. Ex: for the name field, active is now an immutable property.
  - Changed item api to filter out metadata fields that are inactive.
  - Changed search to filter out inactive fields from facets and sorts.

### CODE REVIEW

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js [hunk 1]
```diff
@@ -6,6 +6,47 @@ var APIResponseMixin = require("../mixins/APIResponseMixin");
 var update = require("react-addons-update");
 
 class MetaDataConfigurationActions extends NodeEventEmitter {
+  changeActive(fieldName, activeValue, pushToUrl){
+    // Clone values in order to revert the store if the change fails
+    const previousValue = MetaDataConfigurationStore.fields[fieldName].active;
```
Comment (jgondron) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: False

File: app/assets/javascripts/stores/MetaDataConfigurationStore.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ class MetaDataConfigurationStore extends EventEmitter {
     this._data = {};
 
     Object.defineProperty(this, "fields", { get: function() { return this._data.fields; } });
+    Object.defineProperty(this, "activeFields", { get: function() { return _.where(this._data.fields, {active: true}); } });
```
Comment (jgondron) [inline app/assets/javascripts/stores/MetaDataConfigurationStore.js]: False

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js [hunk 3]
```diff
@@ -6,6 +6,47 @@ var APIResponseMixin = require("../mixins/APIResponseMixin");
...
+    var fieldValues = update(MetaDataConfigurationStore.fields[fieldName], {});
```
Comment (jgondron) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: False

**Comment by jonhartzler** [comment]
This looks good but I am not seeing specs on the changes in sort.rb, search/facet.rb or configuration/facet.rb

**Comment by jonhartzler** [comment]
:+1:

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js (hunk 1 above)
Comment (rfox2) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: Yeah, why does hound do this? I've had this happen to me a few times and gone "huh? - it's defined right there!"

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js [hunk 4]
```diff
@@ -6,6 +6,47 @@ var APIResponseMixin = require("../mixins/APIResponseMixin");
...
+      actionType: MetaDataConfigurationActionTypes.MDC_CHANGE_FIELD,
+      name: fieldName,
+      values: fieldValues,
+    });
+
+    pushToUrl += "/" + fieldName;
+    $.ajax({
+      url: pushToUrl,
+      dataType: "json",
+      method: "PUT",
```
Comment (rfox2) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: Is this component only responsible for updating existing metadata field values?

File: app/assets/javascripts/components/forms/ItemMetaDataForm.jsx [hunk 5]
```diff
@@ -39,7 +39,7 @@ var ItemMetaDataForm = React.createClass({
 
   getInitialState: function() {
     return {
-      formFields: _.sortBy(MetaDataConfigurationStore.fields, 'order'),
+      formFields: _.sortBy(MetaDataConfigurationStore.activeFields, 'order'),
```
Comment (rfox2) [inline app/assets/javascripts/components/forms/ItemMetaDataForm.jsx]: Are the active fields the fields that they have "switched on"?

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js (hunk 4 above)
Comment (jgondron) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: This method only changes the active property for existing fields, but the component as a whole updates and creates. The changeField method allows creating new fields

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx [hunk 6]
```diff
@@ -1,6 +1,18 @@
-var React = require('react');
+var React = require("react");
 var mui = require("material-ui");
 
+var ReactLink = require('react/lib/ReactLink');
+var ReactStateSetters = require('react/lib/ReactStateSetters');
+var MetaDataConfigurationActions = require("../../actions/MetaDataConfigurationActions");
+var Paper = mui.Paper;
```
Comment (rfox2) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: So, this is just a gripe but I wish we didn't have to explicitly include the MUI components that the component uses. I can see arguments for either way: explicit inclusion or implicit but if we're using MUI it sort of seems like it's OK for that to be implicit.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx [hunk 7]
```diff
@@ -8,23 +20,30 @@ var MetaDataConfigurationForm = React.createClass({
...
     return {
-      fields: this.sortedFields(),
+      fields: this.filteredFields(false),
       selectedField: undefined,
+      showInactive: false,
     };
   },
 
-  componentWillMount: function() {
+  componentDidMount: function() {
```
Comment (rfox2) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: This is using Flux right? So what you're saying here is that after the component mounts, check the MetaDataConfigurationStoreChanged attr on the flux store? Just trying to understand the flow.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx [hunk 8]
```diff
@@ -41,25 +60,108 @@ var MetaDataConfigurationForm = React.createClass({
     }
   },
 
-  getFieldRows: function() {
+  getLeftIcon: function(type) {
+    switch(type){
```
Comment (rfox2) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: Where's the "are you some sort of crazy academic for including this??" icon? I think we need that.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx [hunk 9]
```diff
@@ -41,25 +60,108 @@ var MetaDataConfigurationForm = React.createClass({
...
+  backgroundStyle: function() {
+    return {
+      maxWidth: "400px",
+      marginTop: "0px",
+      marginLeft: "48px",
+      padding: "20px"
+    };
+  },
+
+  listStyle: function() {
```
Comment (rfox2) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: This seems crazy. You're always going to return nil right?

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx [hunk 10]
```diff
@@ -69,24 +171,14 @@ var MetaDataConfigurationForm = React.createClass({
   render: function(){
     const { selectedField } = this.state;
     return (
-      <div>
+      <Paper style={ this.backgroundStyle() } zDepth={0}>
```
Comment (rfox2) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: Why did you need the zDepth attr?

File: app/assets/javascripts/stores/MetaDataConfigurationStore.js (hunk 2 above)
Comment (rfox2) [inline app/assets/javascripts/stores/MetaDataConfigurationStore.js]: If you don't define "_" explicitly, I reject all of the new code you've written here. All of it.

File: app/models/metadata/configuration/field.rb [hunk 11]
```diff
@@ -68,8 +75,14 @@ def update(new_attributes)
         end
         convert_json_to_ruby_keys!(new_attributes)
 
+        if new_attributes.has_key?(:immutable)
```
Comment (rfox2) [inline app/models/metadata/configuration/field.rb]: Why would an attr be "immutable"?

File: app/models/waggle/search/result.rb [hunk 12]
```diff
@@ -24,14 +24,16 @@ def total
       end
 
       def facets
-        adapter_result.facets
+        adapter_result.facets.select(&:active)
```
Comment (rfox2) [inline app/models/waggle/search/result.rb]: Could you explain this syntax to me? I'm not familiar with the ampersand syntax.

**Comment by rfox2** [comment]
Questions

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js (hunk 4 above)
Comment (rfox2) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: Did you see the way John wrote a ternary to determine which HTTP verb should be used depending on the action being taken? For REST, POST is used for create operations and PUT for update operations.

File: app/assets/javascripts/components/forms/ItemMetaDataForm.jsx (hunk 5 above)
Comment (jgondron) [inline app/assets/javascripts/components/forms/ItemMetaDataForm.jsx]: Technically at the moment, it's fields they haven't switched off. Since we're doing a soft delete of meta fields, we have to filter those out.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx (hunk 6 above)
Comment (jgondron) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: In general, we've moved to explicit inclusion for libraries since these components are meant to require everything they need to operate (similar to the css discussion).

That being said, as far as these specific declarations (such as Paper = mui.Paper), it isn't required. Within the renders, I could just as easily have done <mui.Paper /> instead of making this local variable that references mui.Paper. It's a matter of taste I suppose, the render looks cleaner but your declarations at the top are a little more verbose. There are some that I've found do not come with the default material-ui require. All of those have explicit require's to the appropriate library within material-ui.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx (hunk 7 above)
Comment (jgondron) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: It is using Flux. "MetaDataConfigurationStoreChanged" is an event message that the store will emit anytime its data changes. This component is subscribing to that event so that it can update it's state anytime the store changes. In this case, if the store ever changes, this component needs to update its state based on the current list of active fields defined.

The getAll() method is what initializes the store. This technically should not be handled inside this component, but rather through an action. I just haven't gotten around to refactoring that part.

Comment (jgondron) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: Also, something to keep in mind is that flux is really just a design pattern. The idea is you don't directly manipulate a store (which is just data shared between multiple components). Instead you go through an actions object that makes changes to the backend (if needed) and the local store. Your components subscribe to events emitted by the store so that they can update their state when a change that they care about is made to the store. The actual library used here is just a standard node event emitter.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx (hunk 9 above)
Comment (jgondron) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: Yeah it was a testing thing. I have some other changes that need to be made to layout to account for the "New field" button, so I'll pull this out as part of those changes.

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx (hunk 10 above)
Comment (jgondron) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: Paper by default has a depth of 1 to always give it a small shadow. It doesn't look right on this subform system we have in settings.

File: app/models/metadata/configuration/field.rb (hunk 11 above)
Comment (jgondron) [inline app/models/metadata/configuration/field.rb]: Ex: The name field should never be allowed to be removed since our application uses it heavily. Some properties of a field like this need to be protected from user changes.

File: app/models/waggle/search/result.rb (hunk 12 above)
Comment (jgondron) [inline app/models/waggle/search/result.rb]: Nor was I. It's filtering to array elements whose active property is true. It's a shorthand version that was suggested by my wise and learned lud (aka the hound).

http://ruby-doc.org/core-1.8.7/Symbol.html#method-i-to_proc

File: app/assets/javascripts/actions/MetaDataConfigurationActions.js (hunk 4 above)
Comment (jgondron) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: Right, which is why this method only does a PUT since it only changes existing fields.

Comment (rfox2) [inline app/assets/javascripts/actions/MetaDataConfigurationActions.js]: Oh sorry. I missed the other ajax call with the ternary. My bad.

File: app/assets/javascripts/components/forms/ItemMetaDataForm.jsx (hunk 5 above)
Comment (rfox2) [inline app/assets/javascripts/components/forms/ItemMetaDataForm.jsx]: OK

File: app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx (hunk 7 above)
Comment (rfox2) [inline app/assets/javascripts/components/forms/MetadataConfigurationForm.jsx]: Gotcha. That's very helpful.

File: app/models/metadata/configuration/field.rb (hunk 11 above)
Comment (rfox2) [inline app/models/metadata/configuration/field.rb]: Ah, I see

File: app/models/waggle/search/result.rb (hunk 12 above)
Comment (rfox2) [inline app/models/waggle/search/result.rb]: Oh, wild. OK.

**Comment by rfox2** [comment]
:+1:
